src/Microsoft.Identity.Web/DownstreamWebApiSupport/DownstreamWebApi.cs

```diff
@@ -106,28 +118,37 @@ public class DownstreamWebApi : IDownstreamWebApi
             TInput input,
             string? authenticationScheme = null,
             Action<DownstreamWebApiOptions>? downstreamWebApiOptionsOverride = null,
-            ClaimsPrincipal? user = null)
+            ClaimsPrincipal? user = null,
+            CancellationToken cancellationToken = default)
             where TOutput : class
         {
             HttpResponseMessage response = await CallWebApiForUserAsync(
                 serviceName,
                 authenticationScheme,
                 downstreamWebApiOptionsOverride,
                 user,
-                new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json")).ConfigureAwait(false);
+                new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json"),
+                cancellationToken).ConfigureAwait(false);
 
             try
             {
                 response.EnsureSuccessStatusCode();
             }
             catch
             {
+#if DOTNET_50_AND_ABOVE
+                string error = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+#else
                 string error = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-
+#endif
```

Comment:
Not quite sure about this `#if` stuff. I've never used it before but I think it's right? `DOTNET_50_AND_ABOVE` is defined in the `Microsoft.Identity.Web.csproj`.
The `ReadAsStringAsync` method didn't take a `cancellationToken`, so I assumed it was because it targeted multiple frameworks. The [5.0 branch of `System.Net.Http`](https://github.com/dotnet/runtime/blob/release/5.0/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L155) includes a `HttpContent.ReadAsStringAsync` that takes a `cancellationToken`.

Reply:
Yeah, this seems to be the way forward, the API was only added in .NET 5 https://stackoverflow.com/questions/25219287/net-httpclient-cancel-readasstringasync

`DOTNET_50_AND_ABOVE` is defined implicity by .NET